class13/slides.mkd

```diff
@@ -0,0 +1,609 @@
+###Application Deployment
+
+@@
+
+###Mark Lavin
+
+* Technical Director at Caktus Group  
+* Co-Author of "Lightweight Django" by O'Reilly Media
+
+@@
+
+###Recap
+
+"New School" Project nearing feature completeness and ready to show to some users
+
+https://github.com/calebsmith/j583-newschool
+
+@@
+
+###Today
+
+* Provision a fresh server for this application
+* Deploy some small changes
+
+
+###Provision
+
+Initially configure servers with current application version
+
+
+###Deploy
+
+Update existing servers with latest application changes
+
+@@
+
+###Web Application Layers
+
+* 1 or more load balancers
+* 1 or more database servers
+* 1 or more application servers
+
+Notes:
+These could be separate physical or virtual servers or all on one machine. Ideal
+configurations scale down to a single development machine and up to a cluster of
+machines. Other pieces like caching, etc may come later.
+
+
+###Our Stack
+
+* 1 load balancer - Nginx
+* 1 database server - SQLite
+* 1 application server - Guniorn
+
+Notes:
+This will all run on one server. Might switch to Postgres if there is time. The
+choice of Nginx and Guniorn could be replaced with Apache, uwsgi, etc. I like these
+because they are easy to use and configure.
+
+@@
+
+###Vagrant
+
+https://www.vagrantup.com/
+
+```ruby
+Vagrant.require_version ">= 1.7.0"
+
+Vagrant.configure("2") do |config|
+  config.vm.box = "ubuntu/trusty64"
+  config.vm.network :private_network, ip: "33.33.33.10"
+  config.vm.provider :virtualbox do |vbox|
+    vbox.customize ["modifyvm", :id, "--memory", "1024"]
+  end
+end
+```
+
+Save this as ``Vagrantfile``
+
+Notes:
+Vagrant is a helpful wrapper around VirtualBox and other virtualization providers.
+You may need/want to use 32 bit if your BIOS isn't configured to support 64 bit
+virtualization.
+
+
+Start Server
+```bash
+local $ vagrant up
+```
+
+Notes:
+The first time this runs it can be very slow in particular if it needs to download
+the box. Those are cached for future use.
+
+
+Connect to Server
+```bash
+local $ vagrant ssh
+```
+
+Notes:
+Connects to the server via SSH as the vagrant user who is an admin on the box.
+
+@@
+
+###Application Setup
+
+Create a new user for running the application
+
+```bash
+vagrant $ sudo useradd --create-home --shell /bin/bash newschool
+```
+
+Notes:
+This is an easy step to skip but we aren't going to. The vagrant user is an
+admin (has sudo access) and we want to minimize the permissions of the user
+which will run the application.
+
+
+Install Python and setup a fresh virtual environment
+
+```bash
+vagrant $ sudo apt-get install wget vim python python-dev git-core
+vagrant $ wget https://bootstrap.pypa.io/get-pip.py
+vagrant $ sudo -H python get-pip.py
+vagrant $ sudo -H pip install virtualenv
+vagrant $ sudo su newschool
+newschool@vagrant $ cd
+newschool@vagrant $ virtualenv newschool -p `which python`
```

Comment:
Missed one instance when I changed the virtualenv name to `env`. This should read `virtualenv env -p 'which python'`. I'll make a PR to update if you don't get to it first.